### PR TITLE
Marging Trade (PostBorrowOrder, GetBorrowOrder & PaySingleBorrowOrder)

### PR DESCRIPTION
**Pull request by hammadsajid786**
Marging Trade #47  fix added

Methods Added
https://docs.kucoin.com/#post-borrow-order (PlaceBorrowOrderAsync)
https://docs.kucoin.com/#get-borrow-order (GetBorrowOrderByClientOrderIdAsync)
https://docs.kucoin.com/#repay-a-single-order (RepaySingleBorrowOrderAsync)

### CODE REVIEW

**Review by hammadsajid786 — COMMENTED** [review]
Changes Tested.
Working fine.

File: Kucoin.Net/Interfaces/IKucoinClientSpot.cs [hunk 1]
```diff
@@ -640,5 +640,56 @@ public interface IKucoinClientSpot : IRestClient
...
+        /// </summary>
+        /// <param name="currency">Currency to Borrow e.g USDT etc</param>
+        /// <param name="type">The type of the order (FOK, IOC)</param>
+        /// <param name="quantity">Total size</param>
+        /// <param name="maxRate">The max interest rate. All interest rates are acceptable if this field is left empty</param>
+        /// <param name="term">erm (Unit: Day). All terms are acceptable if this field is left empty. Please note to separate the terms via comma. For example, 7,14,28</param>
+        /// <param name="ct">Cancellation token</param>
+        /// <returns>The id of the new order</returns>
+        Task<WebCallResult<KucoinNewBorrowOrder>> PlaceBorrowOrderAsync(
+           string currency,
```
Comment (JKorf) [inline Kucoin.Net/Interfaces/IKucoinClientSpot.cs:660]: To keep this in line with the rest of the implementations, this should be called `asset`

File: Kucoin.Net/Interfaces/IKucoinClientSpot.cs [hunk 2]
```diff
@@ -640,5 +640,56 @@ public interface IKucoinClientSpot : IRestClient
...
+        #region Borrow & Lend
+
+        /// <summary>
+        /// Places a Borrow order (https://docs.kucoin.com/#post-borrow-order)
+        /// </summary>
+        /// <param name="currency">Currency to Borrow e.g USDT etc</param>
+        /// <param name="type">The type of the order (FOK, IOC)</param>
+        /// <param name="quantity">Total size</param>
+        /// <param name="maxRate">The max interest rate. All interest rates are acceptable if this field is left empty</param>
+        /// <param name="term">erm (Unit: Day). All terms are acceptable if this field is left empty. Please note to separate the terms via comma. For example, 7,14,28</param>
```
Comment (JKorf) [inline Kucoin.Net/Interfaces/IKucoinClientSpot.cs:656]: Typo `erm` -> `term`

File: Kucoin.Net/SubClients/KucoinClientSpot.cs [hunk 3]
```diff
@@ -1164,6 +1164,81 @@ public async Task<WebCallResult<IEnumerable<KucoinOrder>>> GetStopOrderByClientO
...
+            return await Execute<KucoinNewBorrowOrder>(GetUri("margin/borrow"), HttpMethod.Post, ct, parameters, true).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Get info on a specific borrow order (https://docs.kucoin.com/#get-borrow-order)
+        /// </summary>
+        /// <param name="clientOrderId">The client order id of the borrow order</param>
+        /// <param name="ct">Cancellation token</param>
+        /// <returns>Borrow Order info</returns>
+        public async Task<WebCallResult<KucoinBorrowOrder>> GetBorrowOrderByClientOrderIdAsync(string clientOrderId, CancellationToken ct = default)
```
Comment (JKorf) [inline Kucoin.Net/SubClients/KucoinClientSpot.cs:1207]: I don't think this is actually by `clientOrderId`. By `clientOrderId` would be when you can retrieve the order by an id you passed in when placing the order. This is a generated id, so this should just be called `GetBorrowOrderAsync` and the parameter should be `orderId`

File: Kucoin.Net/Interfaces/IKucoinClientSpot.cs [hunk 4]
```diff
@@ -640,5 +640,56 @@ public interface IKucoinClientSpot : IRestClient
...
+        /// </summary>
+        /// <param name="clientOrderId">The client order id of the borrow order</param>
+        /// <param name="ct">Cancellation token</param>
+        /// <returns>Borrow Order info</returns>
+        Task<WebCallResult<KucoinBorrowOrder>> GetBorrowOrderByClientOrderIdAsync(string clientOrderId, CancellationToken ct = default);
+
+        /// <summary>
+        /// Repay a Single Order (https://docs.kucoin.com/#repay-a-single-order)
+        /// </summary>
+        /// <param name="currency">Currency to Pay e.g USDT etc</param>
```
Comment (JKorf) [inline Kucoin.Net/Interfaces/IKucoinClientSpot.cs:678]: `asset`

File: Kucoin.Net/Interfaces/IKucoinClientSpot.cs [hunk 5]
```diff
@@ -640,5 +640,56 @@ public interface IKucoinClientSpot : IRestClient
...
+        /// <param name="tradeId">Trade ID of borrow order</param>
+        /// <param name="quantity">Repayment size</param>
+        /// <param name="ct">Cancellation token</param>
+        /// <returns>The id of the new order</returns>
```
Comment (JKorf) [inline Kucoin.Net/Interfaces/IKucoinClientSpot.cs:682]: Doesn't return anything

File: Kucoin.Net/Objects/Spot/KucoinBorrowOrder.cs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    {
+        /// <summary>
+        /// The OrderId of the borrow order
+        /// </summary>
+        public string OrderId { get; set; } =  string.Empty;
+
+        /// <summary>
+        /// The currency of the borrow order
+        /// </summary>
+        public string Currency { get; set; } = string.Empty;
```
Comment (JKorf) [inline Kucoin.Net/Objects/Spot/KucoinBorrowOrder.cs:23]: `Asset`

File: Kucoin.Net/Objects/Spot/KucoinBorrowOrder.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        /// <summary>
+        /// The list of Borrow Order Details
+        /// </summary>
+        [JsonProperty("matchList")]
+        public IEnumerable<KucoinBorrowOrderDetails> ExecutionDetails { get; set; } = Array.Empty<KucoinBorrowOrderDetails>();
+
+        /// <summary>
+        /// Status of request
+        /// </summary>
+        public string Status { get; set; } = string.Empty;
```
Comment (JKorf) [inline Kucoin.Net/Objects/Spot/KucoinBorrowOrder.cs:45]: Should be an enum with values `Done` and `Processing`

File: Kucoin.Net/Objects/Spot/KucoinBorrowOrder.cs [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+    /// <summary>
+    /// Kucoin Borrow Execution details
+    /// </summary>
+    public class KucoinBorrowOrderDetails
+    {
+        /// <summary>
+        /// The currency of the borrow order
+        /// </summary>
+        public string Currency { get; set; } = string.Empty;
```
Comment (JKorf) [inline Kucoin.Net/Objects/Spot/KucoinBorrowOrder.cs:56]: `Asset`

File: Kucoin.Net/Objects/Spot/KucoinBorrowOrder.cs [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+        /// <summary>
+        /// Daily interest rate
+        /// </summary>
+        public string DailyIntRate { get; set; } = string.Empty;
```
Comment (JKorf) [inline Kucoin.Net/Objects/Spot/KucoinBorrowOrder.cs:61]: `decimal`

File: Kucoin.Net/Objects/Spot/KucoinBorrowOrder.cs [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+        /// <summary>
+        /// Total size of the borrow order
+        /// </summary>
+        [JsonProperty("size")]
+        public decimal Quantity { get; set; }
+
+        /// <summary>
+        /// Total size of the borrow order
```
Comment (JKorf) [inline Kucoin.Net/Objects/Spot/KucoinBorrowOrder.cs:70]: Incorrect summary

File: Kucoin.Net/Objects/Spot/KucoinBorrowOrder.cs [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        /// </summary>
+        public string Term { get; set; } = string.Empty;
+
+        /// <summary>
+        /// The timestamp of the data
+        /// </summary>
+        [JsonProperty("timestamp"), JsonConverter(typeof(TimestampNanoSecondsConverter))]
```
Comment (JKorf) [inline Kucoin.Net/Objects/Spot/KucoinBorrowOrder.cs:77]: According to the docs this is in milliseconds, not nanoseconds

File: Kucoin.Net/Objects/Spot/KucoinNewBorrowOrder.cs [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        /// <summary>
+        /// The id of the new borrow order
+        /// </summary>
+        public string OrderId { get; set; } = string.Empty;
+
+        /// <summary>
+        /// Currency of current Order
+        /// </summary>
+        public string Currency { get; set; } = string.Empty;
```
Comment (JKorf) [inline Kucoin.Net/Objects/Spot/KucoinNewBorrowOrder.cs:18]: `Asset`

**Review by JKorf — CHANGES_REQUESTED** [review]
Thanks for you PR, please have a look at the comments

**Review by JKorf — APPROVED** [review]

**Pull request merged**